### PR TITLE
Guard against updates with no type changes

### PR DESCRIPTION
**Pull request by davidwengier**
Fixes https://github.com/dotnet/roslyn/issues/56431

### CODE REVIEW

File: src/EditorFeatures/Test/EditAndContinue/EditAndContinueWorkspaceServiceTests.cs [hunk 1]
```diff
@@ -1750,6 +1750,62 @@ public async Task Capabilities()
...
+    // a change that won't cause a type to be emitted
+    int M()
+    {
+        return 1;
+    }
+}
+*/
+";
+
+            var generator = new TestSourceGenerator() { ExecuteImpl = GenerateSource };
```
Comment (tmat) [inline src/EditorFeatures/Test/EditAndContinue/EditAndContinueWorkspaceServiceTests.cs:1781]: Is the usage of code generators necessary to repro?

**Review by tmat — APPROVED** [review]

File: src/EditorFeatures/Test/EditAndContinue/EditAndContinueWorkspaceServiceTests.cs (hunk 1 above)
Comment (davidwengier) [reply src/EditorFeatures/Test/EditAndContinue/EditAndContinueWorkspaceServiceTests.cs:1781]: It was the only case in our existing battery of tests that resulted in an empty updated types list, but there could be other repros, I didn't look too hard.

Comment (tmat) [reply src/EditorFeatures/Test/EditAndContinue/EditAndContinueWorkspaceServiceTests.cs:1781]: Wouldn't just adding the comment directly (w/o source generator) do?

Comment (davidwengier) [reply src/EditorFeatures/Test/EditAndContinue/EditAndContinueWorkspaceServiceTests.cs:1781]: Yes, it does. I could have sworn I tried it, but maybe I had the comment in the wrong place ¯\\\_(ツ)_/¯

File: src/EditorFeatures/Test/EditAndContinue/EditAndContinueWorkspaceServiceTests.cs [hunk 2]
```diff
@@ -1750,6 +1750,62 @@ public async Task Capabilities()
...
+            // validate solution update status and emit
+            var (updates, emitDiagnostics) = await EmitSolutionUpdateAsync(debuggingSession, solution);
+            Assert.Empty(emitDiagnostics);
+            Assert.Equal(ManagedModuleUpdateStatus.Ready, updates.Status);
+
+            // check that no types have been updated. this used to throw
+            var delta = updates.Updates.Single();
+            Assert.Empty(delta.UpdatedTypes);
+
+            EndDebuggingSession(debuggingSession);
```
Comment (MaStr11) [inline src/EditorFeatures/Test/EditAndContinue/EditAndContinueWorkspaceServiceTests.cs:1806]: @davidwengier  This line throws an `System.InvalidOperationException : Pending update has not been committed or discarded.` in the current main branch. See the CI tags in the readme https://github.com/dotnet/roslyn#desktop-unit-tests
The failure is reproducible locally on my machine. A detailed log can be found here https://dev.azure.com/dnceng/public/_build/results?buildId=1377343&view=ms.vss-test-web.build-test-results-tab&runId=40162926&resultId=210614&paneView=debug

Comment (davidwengier) [reply src/EditorFeatures/Test/EditAndContinue/EditAndContinueWorkspaceServiceTests.cs:1806]: Thanks for the heads up!

Fix in https://github.com/dotnet/roslyn/pull/56605

**Pull request merged**